core/src/main/java/io/questdb/cutlass/line/tcp/TableUpdateDetails.java

```diff
@@ -318,16 +320,84 @@ int getColumnCount() {
             return columnCount;
         }
 
+        StringSink getTempSink() {
+            return tempSink;
+        }
+
         int getColumnIndex(DirectByteCharSequence colName) {
-            final int colIndex = columnIndexByNameUtf8.get(colName);
-            if (colIndex != CharSequenceIntHashMap.NO_ENTRY_VALUE) {
+            int colIndex = columnIndexByNameUtf8.get(colName);
+            if (colIndex == COLUMN_INDEX_NOT_CACHED) {
+                colIndex = getColumnIndexFromReader(colName);
+                if (colIndex < 0) {
+                    unresolvedColumnFlag = true;
+                }
+            }
+            return colIndex;
+        }
+
+        private int getColumnIndexFromReader(DirectByteCharSequence colName) {
+            try (TableReader reader = engine.getReader(AllowAllCairoSecurityContext.INSTANCE, tableNameUtf16)) {
+                TableReaderMetadata metadata = reader.getMetadata();
+                tempSink.clear();
+                if (!Chars.utf8Decode(colName.getLo(), colName.getHi(), tempSink)) {
+                    throw CairoException.instance(0).put("invalid UTF8 in value for ").put(colName);
+                }
+                int colIndex = metadata.getColumnIndexQuiet(tempSink);
+                columnIndexByNameUtf8.put(colName.toString(), colIndex);
+                updateColumTypeCache(metadata, colIndex);
                 return colIndex;
             }
-            return populateCacheAndGetColumnIndex(colName);
+        }
+
+        void updateColumnIndexCache() {
+            if (!unresolvedColumnFlag) {
+                return;
+            }
+
+            try (TableReader reader = engine.getReader(AllowAllCairoSecurityContext.INSTANCE, tableNameUtf16)) {
+                TableReaderMetadata metadata = reader.getMetadata();
+                unresolvedColumnFlag = false;
+                boolean resolvedNewColumn = false;
+                Iterator<ObjIntHashMap.Entry<CharSequence>> entries = columnIndexByNameUtf8.iterator();
```

Comment:
we should not be using iterator here, I would perhaps delete `iterator()` method on ObjIntHashMap.

Reply:
iterator has been removed